### PR TITLE
feat(dist): refine suggestions regarding manifest checksum mismatches

### LINKED ISSUE #3885: Confusing error message `toolchain "stable-x86_64-unknown-linux-gnu" is not installable`
### Verification

- [X] I searched for recent similar issues at https://github.com/rust-lang/rustup/issues?q=is%3Aissue+is%3Aopen%2Cclosed and found no duplicates.
- [X] I am on the latest version of Rustup according to https://github.com/rust-lang/rustup/tags and am still able to reproduce my issue.

### Problem

Hi,

My bug occurs when using the command "rustup default stable".

As you may see,  i'm trying to install the latest rust version. My rustup path is changed through a Artifactory path mirroring the official one through a proxy.
What i don't understand is the "version is not installable" as if i don't have the rights or it can't connect to the repository.
Can you please help me ?

### Steps

1. Set up a artifactory mirror to Rust depository
2. Install rustup through rustup-init.sh (seems to be working)
3. use "rustup default stable" or any other version install.

### Possible Solution(s)

As i thought the problem may be two things :

- I didn't properly setup artifactory mirroring (but i can download rustup through .sh so weird)
- I don't have sudo privilegies (can't do anything about it sorry)

### Notes

_No response_

### Rustup version

```console
rustup 1.27.1 (54dd3d00f 2024-04-24)
info: This is the version for the rustup toolchain manager, not the rustc compiler.
```

### Installed toolchains

```console
None
```

### OS version

```console
Redhat 7.3 x86_64
```

### PR DESCRIPTION
**Pull request by rami3l**
Closes #3885.

>  I've made the following update to the error messages nonetheless:
> 
> ### Before
> 
> ```console
>     info: syncing channel updates for 'nightly-aarch64-apple-darwin'
>     info: update not yet available, sorry! try again later
>     error: toolchain 'nightly-aarch64-apple-darwin' is not installable
> ```
> 
> ### After
> 
> ```console
>     info: syncing channel updates for 'nightly-aarch64-apple-darwin'
>     warn: checksum failed for <SNIP>, expected: 'aaaaaaaaaa0f87d2df5e0cff356df307b6649a7b0b4effdf4b6dbd15070206de', calculated: '907f97b96e0f87d2df5e0cff356df307b6649a7b0b4effdf4b6dbd15070206de'
>     info: if you are on the official release server, this is most likely due to an update happening right now, please try again later
>     info: if you are on a third-party release server, this might indicate a problem with the server's configuration
>     error: toolchain 'nightly-aarch64-apple-darwin' is not installable
> ```
> 
> _https://github.com/rust-lang/rustup/issues/3885#issuecomment-2212318125_

### CODE REVIEW

File: src/dist/notifications.rs [hunk 1]
```diff
@@ -150,9 +148,6 @@ impl<'a> Display for Notification<'a> {
                 write!(f, "latest update on {date}, no rust version")
             }
             DownloadingLegacyManifest => write!(f, "manifest not found. trying legacy manifest"),
-            ManifestChecksumFailedHack => {
```
Comment (rami3l) [inline src/dist/notifications.rs:153]: I realized that this notification variant is used only once, so it makes sense to just inline it.

Comment (rami3l) [reply src/dist/notifications.rs:153]: ... in the long run it might even make sense to remove the whole `notification_handler` mechanism?

Comment (djc) [reply src/dist/notifications.rs:153]: Yeah... I think we should unwind a lot of this `Notification` infrastructure.

File: src/dist/mod.rs [hunk 2]
```diff
@@ -1169,9 +1169,18 @@ pub(crate) async fn dl_v2_manifest(
...
+                any.downcast_ref::<RustupError>()
+            {
+                // Manifest checksum mismatched.
+                // This might indicate a transient state during an update
+                // on the official release server
+                // (see <https://github.com/rust-lang/rustup/issues/3390>)
+                // or a misconfiguration of a third-party release server
+                // (see <https://github.com/rust-lang/rustup/issues/3885>).
+                warn!("{err}");
+                info!("if you are on the official release server, this is most likely due to an update happening right now, please try again later");
```
Comment (djc) [inline src/dist/mod.rs]: I think the wording of a "on the [..] server" and "on a [..] server" is a little odd. Also should we limit line lengths on these logging lines to split at better boundaries?

Comment (rami3l) [reply src/dist/mod.rs]: @djc I've slightly adjusted the approach to smartly print one `info!()` line only depending on the current `dist_root_server`. Hopefully, this has also made the notes easier to read.

File: src/dist/mod.rs [hunk 3]
```diff
@@ -1166,9 +1169,20 @@ pub(crate) async fn dl_v2_manifest(
...
+            if let Some(err @ RustupError::ChecksumFailed { .. }) =
+                any.downcast_ref::<RustupError>()
+            {
+                // Manifest checksum mismatched.
+                warn!("{err}");
+
+                let server = dist_root_server(download.process)?;
+                if server == DEFAULT_DIST_SERVER {
+                    // https://github.com/rust-lang/rustup/issues/3390
+                    info!("this is likely due to an ongoing update of the official release server, please try again later");
```
Comment (djc) [inline src/dist/mod.rs:1180]: Maybe add these links to the actual output? At least for I always like the chance to better understand the details of why a problem happens.

**Review by djc — APPROVED** [review]
Nice!

**Pull request merged**